narrow: Resolve the dependency between `narrow` and `stream_list`.

jychen630 opened the pull request:
`narrow.js` and `stream_list.js` forms 2-way cyclic import. To break the cycle, I made `stream_list.initialize()` register  callback functions `handle_narrow_activated` and `handle_narrow_deactivated` that `narrow.js` will use.

On the top of `narrow.js`, a register function is created for receiving the callback function. A gobal variable named similarly to the callback function is created as well. The global variable
`stream_list_handle_narrow_activated` stores the callback and removes the dependency on calling `handle_narrow_activated` in `stream_list.js`. Same for `stream_list_handle_narrow_deactivated` and
`handle_narrow_deactivated`.

This technique is inspired by the dependency breaking method used in #24628.

~~**Current limitation**~~
~~The CI fails at the test case `narrow_activate.test.js` because the mocked `stream_list` and their mocked functions (such as ``) does not call `initialized()`. Therefore, the callback function in `narrow.js` is not registered and thus the test case fails. I tried to `zrequire('stream_list')` and remove `mock_esm("../src/stream_list")` but then the test case shows~~

```
Error:
        Please use mock_template if your test needs to render subscribe_to_more_streams.hbs

        We use mock_template in our unit tests to verify that the
        JS code is calling the template with the proper data. And
        then we use the results of mock_template to supply the JS
        code with either the actual HTML from the template or some
        kind of zjquery stub.
```


~~Since I'm not sure whether following the error message to  mock templates of `stream_list` in `narrow_activate.test.js` is a best practice, so I pause here and would like to pick up later when I get inspired.~~


<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Lalit3716 commented on web/src/stream_list.js:
```diff
@@ -597,6 +597,11 @@ export function initialize() {
+    // register a callback function in narrow.js so that we 
+    // wont need narrow.js to import stream_list.js module
+    // which induce cyclic import.
+    narrow.register_stream_list_callback(handle_narrow_deactivated);
```
It should be `handle_narrow_activated` here instead of `handle_narrow_deactivated`. You have also made this mistake in your commit message.

jychen630 replied:
You're right. Thanks for pointing out!

Lalit3716 commented:
Regarding failing test you could maybe manually register the callback before calling `narrow.activate` in test like so:-
```javascript
    narrow.register_stream_list_callback(stream_list.handle_narrow_activated);
    narrow.activate(terms, {
        then_select_id: selected_id,
    });
```
That will fix the test.

jychen630 commented:
> Regarding failing test you could maybe manually register the callback before calling `narrow.activate` in test like so:-
> 
> ```js
>     narrow.register_stream_list_callback(stream_list.handle_narrow_activated);
>     narrow.activate(terms, {
>         then_select_id: selected_id,
>     });
> ```
> 
> That will fix the test.

Seems to be a workaround. Currently included in my commits.

jychen630 commented:
I have updated the PR using an "ugly" approach from the discussion. Let me know if it is in the right redirection! Converted this to draft because there are still test cases to be updated.

https://chat.zulip.org/#narrow/stream/6-frontend/topic/simple.20import.20cycles/near/1548672

andersk commented on web/src/narrow_hooks.ts:
```diff
@@ -0,0 +1,68 @@
+type Hook = () => void;
+type Hook_ = (param: any) => void;
```
Do not use `any`. This is hiding all the places where you unsafely convert from functions that don’t take arguments to functions that do; also, we never want to disable type checking for the argument.

jychen630 replied:
Yes, thank you for the advice! I understand this totally so this is actually a WIP - just want to see if this is a right direction. I was planning to write a precise type annotaiton later.

timabbott commented:
Closing, as I don't think this is the API we want for breaking the import cycle; and probably some of the refactoring in #23812 is a better next step towards this goal.
